[BUG] up: logPrinter panicked with "send on closed channel"

### Description

This won't be the case every time, but it did show up once.

```
^CGracefully stopping... (press Ctrl+C again to force)
Aborting on container exit...
panic: send on closed channel

goroutine 46 [running]:
github.com/docker/compose/v2/pkg/compose.(*printer).HandleEvent(0xc0007d8330?, {0x0, {0xc00072add1, 0x8}, {0xc0004f22c0, 0x40}, {0xc00072ae00, 0x8}, {0x10319bde8, 0x1}, ...})
        github.com/docker/compose/v2/pkg/compose/printer.go:68 +0x28
github.com/docker/compose/v2/pkg/compose.(*composeService).attachContainer.func1({0x10319bde8?, 0xc0004b9f08?})
        github.com/docker/compose/v2/pkg/compose/attach.go:75 +0xd8
github.com/docker/compose/v2/pkg/utils.(*splitWriter).Close(0xc0005f0780)
        github.com/docker/compose/v2/pkg/utils/writer.go:60 +0x5f
github.com/docker/compose/v2/pkg/compose.(*composeService).attachContainerStreams.func5()
        github.com/docker/compose/v2/pkg/compose/attach.go:152 +0x1ae
created by github.com/docker/compose/v2/pkg/compose.(*composeService).attachContainerStreams in goroutine 1
        github.com/docker/compose/v2/pkg/compose/attach.go:144 +0x385
Exit 2
```

### Steps To Reproduce

1. docker compose up
2. Hit Ctrl-C

You'll need some luck.

### Compose Version

```Text
Docker Compose version v2.23.0-desktop.1
```


### Docker Environment

```Text
Client:
 Version:    24.0.6
 Context:    desktop-linux
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.11.2-desktop.5
    Path:     /.docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.23.0-desktop.1
    Path:     /.docker/cli-plugins/docker-compose
  dev: Docker Dev Environments (Docker Inc.)
    Version:  v0.1.0
    Path:     /.docker/cli-plugins/docker-dev
  extension: Manages Docker extensions (Docker Inc.)
    Version:  v0.2.20
    Path:     /.docker/cli-plugins/docker-extension
  init: Creates Docker-related starter files for your project (Docker Inc.)
    Version:  v0.1.0-beta.9
    Path:     /.docker/cli-plugins/docker-init
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc.)
    Version:  0.6.0
    Path:     /.docker/cli-plugins/docker-sbom
  scan: Docker Scan (Docker Inc.)
    Version:  v0.26.0
    Path:     /.docker/cli-plugins/docker-scan
  scout: Docker Scout (Docker Inc.)
    Version:  v1.0.9
    Path:     /.docker/cli-plugins/docker-scout

Server:
 Containers: 11
  Running: 6
  Paused: 0
  Stopped: 5
 Images: 50
 Server Version: 24.0.6
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc io.containerd.runc.v2
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 8165feabfdfe38c65b599c4993d227328c231fca
 runc version: v1.1.8-0-g82f18fe
 init version: de40ad0
 Security Options:
  seccomp
   Profile: unconfined
  cgroupns
 Kernel Version: 6.4.16-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 15.62GiB
 Name: linuxkit-2a2968bb4c94
 ID: 49b45ad8-9d68-46d3-93b9-22feb4889207
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5555
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: daemon is not using the default seccomp profile
```


### Anything else?

The atomic bool `stopped` (seems to) guards against the writes to the channel:
https://github.com/docker/compose/blob/e4fb5545f02e859b4de8031d927204973d84e67c/pkg/compose/printer.go#L62-L69
...but there's still a possibility that `Stop()` will race with them:

https://github.com/docker/compose/blob/e4fb5545f02e859b4de8031d927204973d84e67c/pkg/compose/printer.go#L55-L60
A timing issue exists between line 67 and line 68. If `Stop()` races with the writes, it could result in a closed channel.

I've attempted to fix this problem [here](https://gist.github.com/horus/908c41c8c9fca7e3fd872dfa053db399), please take a look.

Thank you!

--- a/pkg/compose/printer.go
+++ b/pkg/compose/printer.go
@@ -18,7 +18,7 @@ package compose
 
 import (
 	"fmt"
-	"sync/atomic"
+	"sync"
 
 	"github.com/docker/compose/v2/pkg/api"
 )
@@ -32,9 +32,10 @@ type logPrinter interface {
 }
 
 type printer struct {
+	sync.Mutex
 	queue    chan api.ContainerEvent
 	consumer api.LogConsumer
-	stopped  atomic.Bool
+	stopped  bool
 }
 
 // newLogPrinter builds a LogPrinter passing containers logs to LogConsumer
@@ -53,16 +54,21 @@ func (p *printer) Cancel() {
 }
 
 func (p *printer) Stop() {
-	if p.stopped.CompareAndSwap(false, true) {
+	p.Lock()
+	defer p.Unlock()
+	if !p.stopped {
 		// only close if this is the first call to stop
+		p.stopped = true
 		close(p.queue)
 	}
 }
 
 func (p *printer) HandleEvent(event api.ContainerEvent) {
-	// prevent deadlocking, if the printer is done, there's no reader for
-	// queue, so this write could block indefinitely
-	if p.stopped.Load() {
+	p.Lock()
+	defer p.Unlock()
+	if p.stopped {
+		// prevent deadlocking, if the printer is done, there's no reader for
+		// queue, so this write could block indefinitely
 		return
 	}
 	p.queue <- event

